README.md

# glmPredict

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) <!-- badges: end -->

glmPredict extends the functionality of the `augment.glm` function of the `broom` package to make and evaluate class predictions for a dataset with a Generalised Linear Model. This package was created for an assignment in [STAT545B](https://stat545.stat.ubc.ca/).

## Installation

You can install the development version from [GitHub](https://github.com/janetxinli/glmPredict) with:

``` r
install.packages("devtools")
devtools::install_github("janetxinli/glmPredict")
```

## Examples

We're going to predict the malignancy of samples in a cleaned version of the [Breast Cancer Wisconsin Dataset](https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)) by creating a simple Generalised Linear Model. The dataset, `cancer_clean`, is available in the `glmPredict` package, where it has been slightly shrunk down from the original. The final diagnosis of the samples are encoded by the `malignant` column, where a value of 1 means the sample was diagnosed as malignant, and a sample of 0 means the sample was benign.

First, we need to make the `glm`, with `family=binomial` or `family=quasibinomial` to specify the response as a factor. The data will be subsetted as well, so we have a set of unseen data to pass into `glm_predict` later.

``` r
library(glmPredict)

# Start by subsetting the data
set.seed(123)
subset_len <- floor(0.75 * nrow(cancer_clean))
subset_ind <- sample(seq_len(nrow(cancer_clean)), size=subset_len)
cancer_subset <- cancer_clean[subset_ind,]
cancer_new <- cancer_clean[-subset_ind,]

# Create a glm with cancer_subset
(cancer_model <- glm(malignant ~ texture_mean + radius_mean, data=cancer_subset, family="binomial"))
#> 
#> Call:  glm(formula = malignant ~ texture_mean + radius_mean, family = "binomial", 
#>     data = cancer_subset)
#> 
#> Coefficients:
#>  (Intercept)  texture_mean   radius_mean  
#>     -19.0491        0.2094        1.0021  
#> 
#> Degrees of Freedom: 425 Total (i.e. Null);  423 Residual
#> Null Deviance:       551.5 
#> Residual Deviance: 219.8     AIC: 225.8
```

Then, we can use the `cancer_model` to make predictions, setting the threshold for predicting the positive class (malignant in this case) to 0.6 for example purposes:

``` r
(cancer_predictions <- glm_predict(cancer_model, "malignant", threshold=0.6))
#> # A tibble: 426 x 11
#>    .rownames malignant texture_mean radius_mean .fitted  .resid .std.resid
#>    <chr>         <dbl>        <dbl>       <dbl>   <dbl>   <dbl>      <dbl>
#>  1 415               1         29.8       15.1  9.13e-1  0.425      0.429 
#>  2 463               0         27.0       14.4  7.38e-1 -1.64      -1.66  
#>  3 179               0         22.2       13.0  2.05e-1 -0.677     -0.680 
#>  4 526               0         13.1        8.57 4.45e-4 -0.0299    -0.0299
#>  5 195               1         23.2       14.9  6.69e-1  0.897      0.902 
#>  6 118               1         16.7       14.9  3.42e-1  1.47       1.47  
#>  7 299               0         18.2       14.3  2.78e-1 -0.808     -0.811 
#>  8 229               0         24.0       12.6  2.01e-1 -0.669     -0.674 
#>  9 244               0         23.8       13.8  4.28e-1 -1.06      -1.06  
#> 10 14                1         24.0       15.8  8.64e-1  0.540      0.543 
#> # … with 416 more rows, and 4 more variables: .hat <dbl>, .sigma <dbl>,
#> #   .cooksd <dbl>, .prediction <dbl>
```

By default, `glm_predict` considers the positive class label to be `1` and the negative class to be `0`, which is the standard for binary classification problems, but you can provide different labels using the `pos_label` and `neg_label` arguments. Note that the y value label must still be between 0 and 1, which is also a limitation of the `glm` function.

The function returns the `augment` object, with an additional `.prediction` column containing the predicted class for each observation/example.

``` r
cancer_predictions[[".prediction"]]
#>   [1] 1 1 0 0 1 0 0 0 0 1 1 1 0 0 1 0 0 1 1 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 1 0 1
#>  [38] 0 0 0 0 0 1 1 0 0 0 1 0 1 0 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 0
#>  [75] 0 1 0 0 0 1 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
#> [112] 0 1 0 0 0 0 1 1 0 0 1 0 0 1 0 0 0 0 0 1 0 1 1 0 1 1 0 0 0 1 0 0 0 1 0 0 0
#> [149] 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 0 0 1 0 0 1
#> [186] 0 0 0 1 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1
#> [223] 0 0 1 0 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 1
#> [260] 0 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 0 1 0 0 0 0 1 0 1 0 1 1
#> [297] 0 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 0 0 0
#> [334] 1 0 0 0 1 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0
#> [371] 0 1 1 0 0 0 0 0 1 0 1 0 1 0 1 1 1 1 1 0 1 0 1 1 0 0 1 1 0 0 0 0 0 1 1 0 1
#> [408] 0 1 1 0 0 0 0 1 1 0 0 0 0 0 0 1 0 0 1
```

You can also pass a different dataset (one not used to create the `glm`) to make predictions with `glm_predict`. To do so, you just need to specify the dataset with the `newdata` argument.

``` r
glm_predict(cancer_model, "malignant", threshold=0.6, newdata=cancer_new)
#> # A tibble: 143 x 13
#>    malignant radius_mean texture_mean perimeter_mean area_mean smoothness_mean
#>        <dbl>       <dbl>        <dbl>          <dbl>     <dbl>           <dbl>
#>  1         1        18.0         10.4          123.      1001           0.118 
#>  2         1        20.6         17.8          133.      1326           0.0847
#>  3         1        13           21.8           87.5      520.          0.127 
#>  4         1        13.7         22.6           93.6      578.          0.113 
#>  5         1        14.7         20.1           94.7      684.          0.0987
#>  6         1        16.1         20.7          108.       799.          0.117 
#>  7         0        13.1         15.7           85.6      520           0.108 
#>  8         1        18.6         20.2          122.      1094           0.0944
#>  9         1        11.8         18.7           77.9      441.          0.111 
#> 10         1        16.1         17.9          107        807.          0.104 
#> # … with 133 more rows, and 7 more variables: compactness_mean <dbl>,
#> #   concavity_mean <dbl>, concave_points_mean <dbl>, symmetry_mean <dbl>,
#> #   fractal_dimension_mean <dbl>, .fitted <dbl>, .prediction <dbl>
```

There are several metrics for evaluating predictons in a binary classification problem, such as accuracy, precision and recall. `glmPredict` provides functions to calculate these three scores: `accuracy_score`, `precision_score` and `recall_score`. You can score a set of predictions by providing the true labels and predicted labels to one of these functions, for example:

``` r
accuracy_score(cancer_subset[["malignant"]], cancer_predictions[[".prediction"]])
#> [1] 0.8920188
```

You can also save the true labels and predictions as variables and call one of the scoring functions on the individual vectors:

``` r
true_labels <- cancer_subset[["malignant"]]
predicted_labels <- cancer_predictions[[".prediction"]]
precision_score(true_labels, predicted_labels)
#> [1] 0.9186992
recall_score(true_labels, predicted_labels)
#> [1] 0.7583893
```

## Code of Conduct

Please note that the glmPredict project is released with a [Contributor Code of Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html). By contributing to this project, you agree to abide by its terms.

------------------------------------------------------------------------

### Exercise 1.1: My Development

I started by using devtools to set up the structure of my package:

``` r
library(devtools)
create_package("~/Documents/school/masters/stat545/glmPredict")
```

I set up Git manually through the command line. Then, I opened up the glmPredict directory in Rstudio and created an R file for my function with `use_r("predict")`.

I copied my function from Assignment 1b directly into the created file, R/predict.R, and made some changes that I thought would make it more usable. Since my function has a few dependencies, I used the use\_package function to explicitly tell devtools that these are required:

``` r
use_package("broom")
use_package("dplyr")
use_package("stats")
use_package("datateachr")
use_pipe()
```

I also changed the function calls in `R/predict.R` to explicitly call the functions from their respective packages (i.e. changed `as.formula(f)` to `stats::as.formula(f)`).

Next, I added the documentation skeleton to my file with `Code > Insert Roxygen Skeleton`, and added some information about the function parameters, return value and a couple of example function calls. Then, I added licensing information and aggregated the documentation:

``` r
use_mit_license("Janet Li")
document()
```

At this point, I wanted to check to make sure that I was on the right track, so I used `load_all()` to do a quick manual test of the function. Everything was working as expected, so I ran `check()`... and got an error! Turns out the `@examples` portion of the roxygen2 skeleton actually runs the code, and I was using variable names that didn't exist in my environment.

To ensure that the examples (and my future tests) are able to run, I used `use_data_raw(name="cancer_clean")` to create an R script to clean a dataset to `data-raw/cancer_clean.R`. After adding my code to this file, I ran `usethis::use_data(cancer_clean, overwrite = TRUE)` to add the `.rda` object to the data directory, then I documented the dataset in `R/data.R`.

To add some tests to my package, I used `use_test()` and moved over my tests from Assignment 1b to the `tests/testthat/test-predict.R`. After running the tests manually, I used `test()` to run the automated test. Everything passed!

To create a vignette, I used `use_vignette("glmPredict")`, and added some more detailed documentation to `vignettes/vignette.Rmd`.

Finally, I added a code of conduct and created a package website with:

``` r
use_code_of_conduct()
use_pkgdown()
build_site()
```

To render this README file, I used `build_readme()`.


NEWS.md
## glmPredict v0.2.0
In this minor release, `glmPredict` was refactored to simplify and specify functionalities. The `glm_predict` function was altered to take different positive/negative class labels, and now returns only the `augment` object with corresponding predictions. Evaluating predictions is now addressed by specific functions -- `accuracy_score`, `precision_score` and `recall_score`.

## glmPredict v0.1.1
No code changes here. A more detailed description of the development of `glmPredict` was added to the main `README`.

## glmPredict v0.1.0
This is the first version of glmPredict! It contains a single function, `glm_predict`, that can make predictions given a `glm` object. It also comes packaged with a toy dataset, `cancer_clean`, which can be used to test and play with. 


tests/testthat.R
library(testthat)
library(glmPredict)

test_check("glmPredict")


tests/testthat/test-predict.R
set.seed(123)
train_size <- floor(0.75 * nrow(cancer_clean))
train_ind <- sample(seq_len(nrow(cancer_clean)), size=train_size)

cancer_train <- cancer_clean[train_ind,]
cancer_test <- cancer_clean[-train_ind,]

cancer_model <- glm(malignant ~ radius_mean + concavity_mean, data=cancer_train, family="quasibinomial")

exp_predictions_0.5 <- broom::augment(cancer_model, type.predict="response") %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.5, 1, 0))

exp_accuracy_0.5 <- sum(exp_predictions_0.5==cancer_train$malignant) / length(cancer_train$malignant)
exp_precision_0.5 <- (sum(cancer_train$malignant==1 & exp_predictions_0.5==1)) / ((sum(cancer_train$malignant==0 & exp_predictions_0.5==1)) +
  (sum(cancer_train$malignant==1 & exp_predictions_0.5==1)))
exp_recall_0.5 <- (sum(cancer_train$malignant==1 & exp_predictions_0.5==1)) / ((sum(cancer_train$malignant==1 & exp_predictions_0.5==0)) +
  (sum(cancer_train$malignant==1 & exp_predictions_0.5==1)))

test_that("Outputs the correct values with default parameters", {
  function_output <- glm_predict(cancer_model)
  expect_equal(function_output$accuracy, exp_accuracy_0.5)  # Calculated accuracy
  expect_equal(function_output$precision, exp_precision_0.5)  # Calculated precision
  expect_equal(function_output$recall, exp_recall_0.5)  # Calculated recall
  expect_equal(dplyr::tibble(prediction=function_output$augment$.prediction), exp_predictions_0.5)  # Predictions
})

exp_predictions_0.8_train <- broom::augment(cancer_model, type.predict="response") %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.8, 1, 0))

exp_accuracy_0.8_train <- sum(exp_predictions_0.8_train==cancer_train$malignant) / length(cancer_train$malignant)

test_that("Works with optional parameters on existing data in model", {
  train_0.8_output <- glm_predict(cancer_model, threshold=0.8)
  expect_equal(train_0.8_output$accuracy, exp_accuracy_0.8_train)  # Calculated accuracy
  expect_equal(dplyr::tibble(prediction=train_0.8_output$augment$.prediction), exp_predictions_0.8_train)  # Predictions
})

exp_predictions_0.4_test <- broom::augment(cancer_model, type.predict="response", newdata=cancer_test) %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.4, 1, 0))

exp_precision_0.4_test <- (sum(cancer_test$malignant==1 & exp_predictions_0.4_test==1)) / ((sum(cancer_test$malignant==0 & exp_predictions_0.4_test==1)) +
  (sum(cancer_test$malignant==1 & exp_predictions_0.4_test==1)))

test_that("Works with optional parameters on new data", {
  test_0.4_output <- glm_predict(cancer_model, threshold=0.4, newdata=cancer_test)
  expect_equal(test_0.4_output$precision, exp_precision_0.4_test)  # Precision
  expect_equal(dplyr::tibble(prediction=test_0.4_output$augment$.prediction), exp_predictions_0.4_test)  # Predictions
})

test_that("Throws errors if improper parameter values are given", {
  expect_error(glm_predict(1:20))  # Improper model type
  expect_error(glm_predict(cancer_model, threshold="half"))  # Improper data type for threshold
  expect_error(glm_predict(cancer_model, threshold=1))  # Improper value for threshold
})


tests/testthat/test-glm_predict.R
context("Test glm_predict")

cancer_model <- glm(malignant ~ radius_mean + concavity_mean, data=cancer_subset, family="quasibinomial")

# Expected outputs
exp_predictions_0.5 <- broom::augment(cancer_model, type.predict="response") %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.5, 1, 0))

exp_predictions_subset_0.8 <- broom::augment(cancer_model, type.predict="response") %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.8, 1, 0))

exp_predictions_new_0.4 <- broom::augment(cancer_model, type.predict="response", newdata=cancer_new) %>%
  dplyr::summarize(prediction=ifelse(.fitted > 0.4, 1, 0))

# Tests
test_that("Outputs the correct values with default parameters", {
  function_output <- glm_predict(cancer_model, "malignant")
  expect_equal(dplyr::tibble(prediction=function_output[[".prediction"]]), exp_predictions_0.5)  # Predictions
})

test_that("Works with optional parameters on existing data in model", {
  subset_0.8_output <- glm_predict(cancer_model, "malignant", threshold=0.8)
  expect_equal(dplyr::tibble(prediction=subset_0.8_output[[".prediction"]]), exp_predictions_subset_0.8)  # Predictions
})

test_that("Works with optional parameters on new data", {
  new_0.4_output <- glm_predict(cancer_model, "malignant", threshold=0.4, newdata=cancer_new)
  expect_equal(dplyr::tibble(prediction=new_0.4_output[[".prediction"]]), exp_predictions_new_0.4)  # Predictions
})

test_that("Throws errors if improper parameter values are given", {
  expect_error(glm_predict(1:20, "malignant"))  # Improper model type
  expect_error(glm_predict(cancer_model, "diagnosis", threshold="half"))  # Improper data type for threshold
  expect_error(glm_predict(cancer_model, threshold=1))  # Improper value for threshold
})


R/data.R
#' Cleaned Diagnostic Breast Cancer Data
#'
#' A subset of a dataset containing quantitative features from images of nuclei from biopsies of breast masses from patients at the University
#' of Wisconsin Hospital. Malignancy is encoded as a binary variable (1 = malignant, 0 = benign).
#'
#' @format A tibble with 569 rows and 11 variables:
#' \describe{
#'   \item{malignant}{Whether the observation was diagnosed as malignant or benign}
#'   \item{radius_mean}{Mean radius of nuclei present in sample image}
#'   \item{texture_mean}{Mean texture of nuclei present in sample image}
#'   \item{perimeter_mean}{Mean perimeter of nuclei present in sample image}
#'   \item{area_mean}{Mean area of nuclei present in sample image}
#'   \item{smoothness_mean}{Mean smoothness of nuclei present in sample image}
#'   \item{compactness_mean}{Mean compactness of nuclei present in sample image}
#'   \item{concavity_mean}{Mean concavity of nuclei present in sample image}
#'   \item{concave_points_mean}{Mean concave points of nuclei present in sample image}
#'   \item{symmetry_mean}{Mean symmetry of nuclei present in sample image}
#'   \item{fractal_dimension_mean}{Mean factal dimension of nuclei present in sample image}
#'   }
#' @source \url{https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+(Diagnostic)}
"cancer_clean"


R/glm_predict.R
#' GLM Predict
#'
#' Create a Generalised Linear Model (glm) from a dataset and a given set of variables, then
#' make predictions from the model, either on the original data or a new dataset.
#'
#' @param model Fitted `glm` model object
#' @param target Target name (must be present as a column in the model object). This is the y variable when creating a `glm`.
#' @param pos_class Positive class label. default=`1`. Like the y value in `glm`, this must be between 0 and 1.
#' @param neg_class Negative class label. default=`0`. Like the y value in `glm`, this must be between 0 and 1.
#' @param threshold Probability threshold for predicting positive class. Must be between 0 and 1 (exclusive). default=`0.5`.
#' @param newdata New data to make predictions on. By default, `glm_predict` predicts on the original data used to create the model. default=`NULL`.
#' @param ... Other parameters to be passed into the `augment` function for making predictions.
#'
#' @return Augment object with a column `.prediction` that contains the predicted class for each example.
#' @export
#'
#' @examples require(dplyr)
#' @examples cancer_model <- glm(malignant ~ texture_mean, data=cancer_clean, family="quasibinomial")
#' @examples glm_predict(cancer_model, "malignant")
#' @examples glm_predict(cancer_model, "malignant", threshold=0.8)
glm_predict <- function(model, target, pos_class=1, neg_class=0, threshold=0.5, newdata=NULL, ...) {
  # Check that input types are as expected
  if (!"glm" %in% class(model)) {
    stop("Error: must provide a fitted Generalised Linear Model")
  }

  if (!target %in% colnames(model$data)) {
    stop("Error: target name provided is not present in model data")
  }

  if (typeof(threshold) != "double") {
    stop("Error: probability threshold must be a double")
  } else if (threshold >= 1 | threshold <= 0) {
    stop("Error: probability threshold value must be between 0 and 1")
  }

  if (!is.null(newdata) & (!"data.frame" %in% class(newdata)) & !("tbl" %in% class(newdata))) {
    stop("Error: newdata must be a data frame or tibble")
  }

  if (pos_class == neg_class) {
    stop("Error: positive and negative class labels cannot be identical")
  }

  # Augment existing or new data
  if (is.null(newdata)) {
    a <- broom::augment(model, type.predict="response", ...)
  }
  else {
    a <- broom::augment(model, type.predict="response", newdata=newdata, ...)
  }

  # Predict based on the given threshold
  a <- a %>%
    dplyr::mutate(.prediction=ifelse(a[[".fitted"]] > threshold, pos_class, neg_class))

  # Return augment object and predictions
  return(a)
}


R/recall_score.R
#' Recall Score
#'
#' @param correct True/correct labels.
#' @param predictions Predicted labels.
#' @param pos_class Positive class label. Like the y value in `glm`, this must be between 0 and 1. default=`1`.
#' @param neg_class Negative class label. Like the y value in `glm`, this must be between 0 and 1. default=`0`.
#'
#' @return `recall` (numeric): The recall (true positives / true positives + false negatives) of the predictions
#' @export
#'
#' @examples cancer_model <- glm(malignant ~ texture_mean, data=cancer_clean, family="quasibinomial")
#' @examples cancer_preds <- glm_predict(cancer_model, "malignant")
#' @examples recall <- recall_score(cancer_clean[["malignant"]], cancer_preds[[".prediction"]])
recall_score <- function(correct, predictions, pos_class=1, neg_class=0) {
  if (length(correct) != length(predictions)) {
    stop("Error: correct labels and predictions are not the same length")
  }
  if (length(unique(correct)) != 2) {
    stop("Error: 'correct label' dataset contains more than two labels")
  }
  if (length(unique(predictions)) != 2) {
    stop("Error: predictions contain more than two labels")
  }
  tp <- sum(correct == pos_class & predictions == pos_class)
  fn <- sum(correct == pos_class & predictions == neg_class)
  return(tp / (tp + fn))
}


R/accuracy_score.R
#' Accuracy Score
#'
#' Given a set of predictions from `glm_predict` and the correct data,
#' returns the accuracy of the predictions.
#'
#' @param correct True/correct labels.
#' @param predictions Predicted labels.
#' @param normalize Normalize the accuracy to the number of samples. default = `TRUE`.
#'
#' @return `accuracy` (numeric): If `normalize==TRUE`, this is the fraction of correctly classified examples. Otherwise, returns the number correct predictions.
#' @export
#' @examples cancer_model <- glm(malignant ~ texture_mean, data=cancer_clean, family="quasibinomial")
#' @examples cancer_preds <- glm_predict(cancer_model, "malignant")
#' @examples accuracy <- accuracy_score(cancer_clean[["malignant"]], cancer_preds[[".prediction"]])
accuracy_score <- function(correct, predictions, normalize=TRUE) {  # Change data arg to `correct`
  if (length(correct) != length(predictions)) {
    stop("Error: data and predictions are not the same length")
  }
  accuracy <- mean(correct == predictions)
  if (normalize) {
    return(accuracy)
  }
  else {
    return(accuracy * length(correct))
  }
}


R/precision_score.R
#' Precision Score
#'
#' @param correct True/correct labels.
#' @param predictions Predicted labels.
#' @param pos_class Positive class label. Like the y value in `glm`, this must be between 0 and 1. default=`1`.
#' @param neg_class Negative class label. Like the y value in `glm`, this must be between 0 and 1. default=`0`.
#'
#' @return `precision` (numeric): The precision (true positives / true positives + false positives) of the predictions
#' @export
#'
#' @examples cancer_model <- glm(malignant ~ texture_mean, data=cancer_clean, family="quasibinomial")
#' @examples cancer_preds <- glm_predict(cancer_model, "malignant")
#' @examples precision <- precision_score(cancer_clean[["malignant"]], cancer_preds[[".prediction"]])
precision_score <- function(correct, predictions, pos_class=1, neg_class=0) {
  if (length(correct) != length(predictions)) {
    stop("Error: correct labels and predictions are not the same length")
  }
  if (length(unique(correct)) != 2) {
    stop("Error: 'correct label' dataset contains more than two labels")
  }
  if (length(unique(predictions)) != 2) {
    stop("Error: predictions contain more than two labels")
  }
  tp <- sum(correct == pos_class & predictions == pos_class)
  fp <- sum(correct == neg_class & predictions == pos_class)
  return(tp / (tp + fp))
}


R/predict.R
#' GLM Prediction
#'
#' Create a Generalised Linear Model (glm) from a dataset and a given set of variables, then
#' make predictions from the model, either on the original data or a new dataset.
#'
#' @param model_data dataframe or tibble
#' @param y target variable name/response (given as a character)
#' @param x features/terms (given as a single character or a character vector)
#' @param threshold probability threshold for predicting positive class, default=0.5
#' @param predict_data new data to make predictions on. by default, glm_predict predicts on the original data used to create the model
#' @param ... other parameters to be passed into glm function
#'
#' @return A named list containing:
#' @return $model: fitted model object
#' @return $augment: augment object containing the predictions and statistics
#' @return $accuracy
#' @return $precision
#' @return $recall
#' @export
#'
#' @examples require(dplyr)
#' @examples glm_predict(cancer_clean, "malignant", "texture_mean")
#' @examples glm_predict(cancer_clean, "malignant", c("radius_mean", "concavity_mean"), threshold=0.8)
glm_predict <- function(model_data, y, x, threshold=0.5, predict_data=NA, ...) {
  # Check that input types are as expected
  if ((!"data.frame" %in% class(model_data)) & (!"tbl" %in% class(model_data))) {
    stop("Error: model_data must be a data frame or tibble")
  }
  if (typeof(x) != "character") {
   stop("Error: feature column name(s) must be given as a character or character vector")
  }
  if (typeof(y) != "character") {
    stop("Error: target column name must be given as a character")
  }
  if (typeof(threshold) != "double") {
    stop("Error: probability threshold must be a double")
  }
  if (!is.na(predict_data) & (!"data.frame" %in% class(predict_data)) & !("tbl" %in% class(predict_data))) {
    stop("Error: predict_data must be a data frame or tibble")
  }

  # Check if columns exist in model_data
  if (!y %in% colnames(model_data)) {
    stop("Error: target column name does not exist within model_data")
  }
  if (sum(x %in% colnames(model_data)) != length(x)) {
    stop("Error: one or more feature column names does not exist within model_data")
  }

  # Create the model and augment object
  if (is.vector(x)) {
    x <- paste(x, collapse=" + ")
  }
  f <- paste(y, "~", x)
  model <- stats::glm(stats::as.formula(f), family="binomial", data=model_data, ...)

  if (is.na(predict_data)) {
    a <- broom::augment(model, type.predict="response")
  }
  else {
    a <- broom::augment(model, type.predict="response", newdata=predict_data)
  }

  # Predict based on the given threshold
  a <- a %>%
    dplyr::mutate(.prediction=ifelse(.fitted > threshold, 1, 0))

  # Calculate error types and metrics
  true_pos <- sum(a[[".prediction"]]==1 & a[[y]]==1)
  false_pos <- sum(a[[".prediction"]]==1 & a[[y]]==0)
  false_neg <- sum(a[[".prediction"]]==0 & a[[y]]==1)

  accuracy <- mean(a[[".prediction"]]==a[[y]])
  precision <- true_pos / (true_pos + false_pos)
  recall <- true_pos / (true_pos + false_neg)

  # Return model, model info and metrics
  list("model"=model, "augment"=a, "accuracy"=accuracy, "precision"=precision, "recall"=recall)
}
